### DISCUSSION TITLE
External media authentication with Cloudflare workers

### DISCUSSION THREAD

**Discussion by ananthb** [tinacms/tinacms#4329] [Help]
My site is served by Cloudflare Pages. I have images in a Cloudflare R2 bucket which offers a complete S3 API. I want to use external media with what is basically an S3 bucket. How do I get authentication working with Cloudflare Workers?

Workers will execute JavaScript code that is present in the `/functions` path of my static site, so I can have it respond to API requests at any path. I'm not sure how to adapt `next-tinacms-s3` to do the authentication bit though.

I tried following the docs here: https://tina.io/docs/reference/media/external/s3/, but `createMediaHandler` expects a request and response via some http server, but Cloudflare Pages Functions has a different interface whereby a callback function is registered and then invoked for each request.

__Cloudflare Pages Function__

```js
export function onRequest(context) {
  return new Response("Hello, world!")
}
```

**Comment by wkelly17** [comment]
@ananthb -> I also took a look at this.  Due to Tina originally being developed in Next, it's looking for Node/Next/Express like Request and Response objects instead of web standards ones, otherwise the package would work as is since the api is the same.   Although it may not be an ideal solution for forever just to keep maintained, one option for now that I successfully did was just to copy 90% of the s3 plugin's code (it's just 3 files. About 400 LOC) into your Cloudflare pages project as a custom media store (it's only 3 files).  It's mostly just tweaking the handler's file to accept a standard Web request, and from taht handler (the pages function) you can just return a web standard response.   There are a few other things in that file that are not needed though. For example, this section, https://github.com/tinacms/tinacms/blob/main/packages/next-tinacms-s3/src/handlers.ts#L89-L101.   You can just send a multipart file upload and process it with formData (no need for the multer + promisify).

**Comment by ananthb** [comment]
Yeah I was thinking of hacking up something along the same lines. I don't see any reason why most of the existing code shouldn't work. Do you have your source code modifications somewhere where I can take a look?

**Reply by wkelly17** [reply]
Yeah, let me strip out some cruft and consolidate some of the pieces.

**Reply by wkelly17** [reply]
@ananthb  - https://codesandbox.io/s/stupefied-fog-n5zjpg
I think this is it.  Mostly copied and then tweaked here and there for some r2 stuff and dropping multer etc; Obv this doesn't run, but should be good.  There's some stuff in there about prefixes and media root that I'm not 100% certain on since I didn't author the original s3 stuff, but in my limited tests, this has worked.

**Comment by danallan** [comment]
I approached this by creating shims for the NextApiRequest and NextApiResponse types expected by the media handler. This is still brittle but it lets me use the existing handler code from `next-tinacms-s3` without duplication.

It's worth noting that this is only now possible because the `path` dependency has been polyfilled in a recent CloudFlare worker runtime compatibility update. The dependency is [imported here](https://github.com/tinacms/tinacms/blob/3bb00d6ef3cb735dd29db644e8e634a878e4d7a4/packages/next-tinacms-s3/src/handlers.ts#L19) in `next-tinacms-s3`, and previously the CloudFlare Worker did not support `path` at all (only `node:path`).

The change is available in [this gist](https://gist.github.com/danallan/c9ead81195a115b64706d2453cc31acf). `createMediaHandler()` is the same as from the docs (including the authorization check). All this does is just create fake request and response objects that include the methods and properties that `next-tinacms-s3` and `isAuthorized()` from `@tinacms/auth` expect. So it could be brittle in the future if any changes to these packages depend on additional features from the NextApi types that are not modeled here, but hopefully it's easy enough to adapt.

The gist is a direct copy from my working project with Astro (v5) with the CloudFlare adapter and using Backblaze B2 which is another S3-compatible API.

To use this:
1. Set the `nodejs_compat` compatibility flag with a compatibility date of `2024-09-23` or later for your Worker, see [CloudFlare docs](https://developers.cloudflare.com/workers/runtime-apis/nodejs/#get-started).

2. Use the contents of [the gist](https://gist.github.com/danallan/c9ead81195a115b64706d2453cc31acf) in your project. You may need to adjust how `req.query.media` is computed depending on your framework, though I think it's only used in DELETEs.

3. Everything else is the same as the [docs](https://tina.io/docs/reference/media/external/s3); add `next-tinacms-s3` and `@tinacms/auth` and register the media store in config. Don't forget to set your CORS settings for your S3-equivalent bucket so that the pre-signed upload URL will work.

If you're still using the copied version discussed previously, make sure that you are guarding access with `isAuthorized()` so you don't have unauthenticated requests against your bucket.
